### DISCUSSION TITLE
You need to have media library pro installed to make this work

### DISCUSSION THREAD

**Discussion by glupeksha** [spatie/laravel-medialibrary#2161] [Q&A] [answered]
Im following the spatie media library documentation (https://spatie.be/docs/laravel-medialibrary/v8/basic-usage/retrieving-media)  and when im trying to retrieve media models I get the following error.
```
Spatie\MediaLibrary\MediaCollections\Exceptions\FunctionalityNotAvailable
You need to have media library pro installed to make this work. 
```
Is it a bug? Since it is in the basic usage section I thought it is available for public.

**Comment by freekmurze** [comment, marked as answer]
Could you post the entire stacktrace?

Could you post the code that triggers the error?

**Reply by glupeksha** [reply]
I tried in tinker and it worked. The issue occurs when im trying to pass data to the component
`<x-form.file name="images" :values="isset($catelogItem)?$catelogItem->media:null" :disable="$disable" />`

**Reply by freekmurze** [reply]
To me it makes no sense that it works in tinker and not anywhere else. It also seems that our testsuite is indicating that all functionality is working properly. Could you PR a failing test for me to look at?

**Reply by glupeksha** [reply]
Stack Trace:

Facade\Ignition\Exceptions\ViewException: You need to have media library pro installed to make this work. (View: D:\Dawn\ERP_v8_2\ERP\resources\views\crud\catelog-items\base.blade.php) in file D:\Dawn\ERP_v8_2\ERP\vendor\spatie\laravel-medialibrary\src\MediaCollections\Exceptions\FunctionalityNotAvailable.php on line 11

#0 D:\Dawn\ERP_v8_2\ERP\vendor\spatie\laravel-medialibrary\src\Support\MediaLibraryPro.php(13): Spatie\MediaLibrary\MediaCollections\Exceptions\FunctionalityNotAvailable::mediaLibraryProRequired()
#1 D:\Dawn\ERP_v8_2\ERP\vendor\spatie\laravel-medialibrary\src\MediaCollections\Models\Collections\MediaCollection.php(37): Spatie\MediaLibrary\Support\MediaLibraryPro::ensureInstalled()
#2 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Support\helpers.php(115): Spatie\MediaLibrary\MediaCollections\Models\Collections\MediaCollection-&gt;toHtml()
#3 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Compilers\Concerns\CompilesComponents.php(174): e(Object(Spatie\MediaLibrary\MediaCollections\Models\Collections\MediaCollection))
#4 D:\Dawn\ERP_v8_2\ERP\resources\views/crud/catelog-items/base.blade.php(81): Illuminate\View\Compilers\BladeCompiler::sanitizeComponentAttribute(Object(Spatie\MediaLibrary\MediaCollections\Models\Collections\MediaCollection))
#5 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(107): require('D:\\Dawn\\ERP_v8_...')
#6 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(108): Illuminate\Filesystem\Filesystem::Illuminate\Filesystem\{closure}()
#7 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\PhpEngine.php(58): Illuminate\Filesystem\Filesystem-&gt;getRequire('D:\\Dawn\\ERP_v8_...', Array)
#8 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\CompilerEngine.php(61): Illuminate\View\Engines\PhpEngine-&gt;evaluatePath('D:\\Dawn\\ERP_v8_...', Array)
#9 D:\Dawn\ERP_v8_2\ERP\vendor\facade\ignition\src\Views\Engines\CompilerEngine.php(37): Illuminate\View\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#10 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(139): Facade\Ignition\Views\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#11 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(122): Illuminate\View\View-&gt;getContents()
#12 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(91): Illuminate\View\View-&gt;renderContents()
#13 D:\Dawn\ERP_v8_2\ERP\resources\views/general/show.blade.php(9): Illuminate\View\View-&gt;render()
#14 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(107): require('D:\\Dawn\\ERP_v8_...')
#15 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(108): Illuminate\Filesystem\Filesystem::Illuminate\Filesystem\{closure}()
#16 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\PhpEngine.php(58): Illuminate\Filesystem\Filesystem-&gt;getRequire('D:\\Dawn\\ERP_v8_...', Array)
#17 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\CompilerEngine.php(61): Illuminate\View\Engines\PhpEngine-&gt;evaluatePath('D:\\Dawn\\ERP_v8_...', Array)
#18 D:\Dawn\ERP_v8_2\ERP\vendor\facade\ignition\src\Views\Engines\CompilerEngine.php(37): Illuminate\View\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#19 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(139): Facade\Ignition\Views\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#20 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(122): Illuminate\View\View-&gt;getContents()
#21 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(91): Illuminate\View\View-&gt;renderContents()
#22 D:\Dawn\ERP_v8_2\ERP\resources\views/crud/catelog-items/show.blade.php(1): Illuminate\View\View-&gt;render()
#23 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(107): require('D:\\Dawn\\ERP_v8_...')
#24 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Filesystem\Filesystem.php(108): Illuminate\Filesystem\Filesystem::Illuminate\Filesystem\{closure}()
#25 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\PhpEngine.php(58): Illuminate\Filesystem\Filesystem-&gt;getRequire('D:\\Dawn\\ERP_v8_...', Array)
#26 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Engines\CompilerEngine.php(61): Illuminate\View\Engines\PhpEngine-&gt;evaluatePath('D:\\Dawn\\ERP_v8_...', Array)
#27 D:\Dawn\ERP_v8_2\ERP\vendor\facade\ignition\src\Views\Engines\CompilerEngine.php(37): Illuminate\View\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#28 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(139): Facade\Ignition\Views\Engines\CompilerEngine-&gt;get('D:\\Dawn\\ERP_v8_...', Array)
#29 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(122): Illuminate\View\View-&gt;getContents()
#30 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\View.php(91): Illuminate\View\View-&gt;renderContents()
#31 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Http\Response.php(62): Illuminate\View\View-&gt;render()
#32 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Http\Response.php(34): Illuminate\Http\Response-&gt;setContent(Object(Illuminate\View\View))
#33 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(765): Illuminate\Http\Response-&gt;__construct(Object(Illuminate\View\View), 200, Array)
#34 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(737): Illuminate\Routing\Router::toResponse(Object(Illuminate\Http\Request), Object(Illuminate\View\View))
#35 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(691): Illuminate\Routing\Router-&gt;prepareResponse(Object(Illuminate\Http\Request), Object(Illuminate\View\View))
#36 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(128): Illuminate\Routing\Router-&gt;Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#37 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Auth\Middleware\Authorize.php(45): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#38 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Auth\Middleware\Authorize-&gt;handle(Object(Illuminate\Http\Request), Object(Closure), 'view', 'catelog_item')
#39 D:\Dawn\ERP_v8_2\ERP\vendor\spatie\laravel-multitenancy\src\Http\Middleware\EnsureValidTenantSession.php(27): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#40 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Spatie\Multitenancy\Http\Middleware\EnsureValidTenantSession-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#41 D:\Dawn\ERP_v8_2\ERP\vendor\spatie\laravel-multitenancy\src\Http\Middleware\NeedsTenant.php(19): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#42 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Spatie\Multitenancy\Http\Middleware\NeedsTenant-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#43 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Middleware\SubstituteBindings.php(41): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#44 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Routing\Middleware\SubstituteBindings-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#45 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Auth\Middleware\Authenticate.php(44): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#46 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Auth\Middleware\Authenticate-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#47 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\VerifyCsrfToken.php(77): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#48 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\VerifyCsrfToken-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#49 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\View\Middleware\ShareErrorsFromSession.php(49): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#50 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\View\Middleware\ShareErrorsFromSession-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#51 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Session\Middleware\StartSession.php(121): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#52 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Session\Middleware\StartSession.php(63): Illuminate\Session\Middleware\StartSession-&gt;handleStatefulRequest(Object(Illuminate\Http\Request), Object(Illuminate\Session\Store), Object(Closure))
#53 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Session\Middleware\StartSession-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#54 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse.php(37): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#55 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#56 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Cookie\Middleware\EncryptCookies.php(67): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#57 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Cookie\Middleware\EncryptCookies-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#58 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(103): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#59 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(693): Illuminate\Pipeline\Pipeline-&gt;then(Object(Closure))
#60 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(668): Illuminate\Routing\Router-&gt;runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#61 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(634): Illuminate\Routing\Router-&gt;runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#62 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Routing\Router.php(623): Illuminate\Routing\Router-&gt;dispatchToRoute(Object(Illuminate\Http\Request))
#63 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(166): Illuminate\Routing\Router-&gt;dispatch(Object(Illuminate\Http\Request))
#64 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(128): Illuminate\Foundation\Http\Kernel-&gt;Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#65 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\TransformsRequest.php(21): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#66 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\TransformsRequest-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#67 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\TransformsRequest.php(21): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#68 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\TransformsRequest-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#69 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\ValidatePostSize.php(27): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#70 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\ValidatePostSize-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#71 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance.php(87): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#72 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#73 D:\Dawn\ERP_v8_2\ERP\vendor\fruitcake\laravel-cors\src\HandleCors.php(37): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#74 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Fruitcake\Cors\HandleCors-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#75 D:\Dawn\ERP_v8_2\ERP\vendor\fideloper\proxy\src\TrustProxies.php(57): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#76 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Fideloper\Proxy\TrustProxies-&gt;handle(Object(Illuminate\Http\Request), Object(Closure))
#77 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(103): Illuminate\Pipeline\Pipeline-&gt;Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#78 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(141): Illuminate\Pipeline\Pipeline-&gt;then(Object(Closure))
#79 D:\Dawn\ERP_v8_2\ERP\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(110): Illuminate\Foundation\Http\Kernel-&gt;sendRequestThroughRouter(Object(Illuminate\Http\Request))
#80 D:\Dawn\ERP_v8_2\ERP\public\index.php(52): Illuminate\Foundation\Http\Kernel-&gt;handle(Object(Illuminate\Http\Request))
#81 D:\Dawn\ERP_v8_2\ERP\server.php(21): require_once('D:\\Dawn\\ERP_v8_...')
#82 {main}

**Reply by glupeksha** [reply]
I think it works in tinker because its not rendering anything to html. But when trying to pass data to a component, this issue triggers.

**Reply by glupeksha** [reply]
**### Failing Code:**
```
//base.blade.php
...
<x-form.file name="images" :values="isset($catelogItem)?$catelogItem->media:null" :disable="$disable" />
...
```
```
//file.blade.php
<x-form.form-input :name="$name">
....
    <div id="{{$name}}_preview" class="d-flex flex-row flex-wrap">
        @foreach($values as $value)
        <div class='p-2'><img class='img-thumbnail' style='max-height:60px' height='60' src='{{$value->getUrl()}}'></div>
        @endforeach
    </div>
...
</x-form.form-input>
```

**### Success Code**
```
//base.blade.php
...
<x-form.file name="images" :values="isset($catelogItem)?$catelogItem:null" :disable="$disable" />
...
```
```
//file.blade.php
<x-form.form-input :name="$name">
...
    <div id="{{$name}}_preview" class="d-flex flex-row flex-wrap">
        @foreach($values->media as $value)
        <div class='p-2'><img class='img-thumbnail' style='max-height:60px' height='60' src='{{$value->getUrl()}}'></div>
        @endforeach
    </div>
...
</x-form.form-input>
```

**Reply by freekmurze** [reply]
Could you try to remove `MediaLibraryPro::ensureInstalled();` at line 37 of the `MediaCollection` class and let me know if that solves the issue?

**Reply by glupeksha** [reply]
It works if I comment that line.. But is it okay to change code in vendor folder?

**Reply by freekmurze** [reply]
I've remove the line in v9.1.1. Run `composer update` to pull that in.

**Reply by glupeksha** [reply]
@freekmurze Thank you so much
